keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+
+A `Catalog` will contain a collection of one or more modules that can be used with a Kustomize resource. 
+
+A minimal example is shown below:
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    provider: 
+      container: 
+        image: example/module_providers/java:v1.0.0
+```
+
+This will enable a users to use plugin providers with provider information omitted, such as:
+
+```yaml
+# app/kustomization.yaml
+kind: Kustomization
+catalogs:
+  - https://example.com/plugins/catalog.yaml
+resources:
+- javaapp.yaml
+```
+
+```yaml
+# app/javaapp.yaml
+apiVersion: example.com/v1
+kind: JavaApplication
+spec:
+  application: team/my-app
+  version: v1.0
+```
+
+When this Kustomization is processed by `kustomize build`, the referenced catalog (or catalogs) will be used to locate a plugin provider that supports the apiVersion `example.com/v1` and kind `JavaApplication`. If found in one of the referenced catalogs, kustomize can determine the provider configuration without the need for the user to specify it in the kustomization resources directly. The catalogs will be searched in order specified. If more than one catalog defines the target apiVersion and kind, the first will be selected. 
+
+In addition to the new catalog kind, `kustomize build` will accept a repeatable flag `--trusted-plugin-catalog=""`. When present, this flag instructs `kustomize build` to automatically fetch and execute plugins that are defined by the catalog and referenced within the Kustomization, Component or Composition. When a resource is processed by `kustomize build` and a catalog is referenced but not specified using the `--trusted-plugin-catalog=""` flag, an error will occur. Kustomize will provide a built in Catalog for supporting official extensions, published to a well publicized endpoint. This catalog will _NOT_ require the user to explicitly trust it. Users can provide the `apiVersion` and `kind` of the official extensions in kustomize resource and these will be resolved by the official catalog.
+
+In addition to container based plugin providers, the `Catalog` will support discovery of Starlark and Exec based plugin providers, via an HTTP(s), Git, or OCI reference as illustrated below: 
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: GroovyApplication
+    description: "A Kustomize plugin provider that can handle groovy apps"
+    provider:  
+      starlark: https://example.co/module_providers/starlark-func:v1.0.0
```

Comment:
I am a fan of not providing built in execution for any particular stack.  The starlark function exists in the kpt catalog: https://catalog.kpt.dev/starlark/v0.2/ .  Execution environments can carry an additional attack and supportability burden.

Reply:
+1

Reply:
Since the over-arching KEP is proposing deprecation of starlark, I think it would be reasonable to update this one to not use Starlark examples.

Reply:
Will do, that makes sense. Do you feel like any of the examples should involve `exec` style or should the focus be containerized?

Reply:
I think we should include at least some exec examples